### PR TITLE
Implement dotnet list projects

### PR DESCRIPTION
**Pull request by jgoshi**
@piotrpMSFT @livarcocc @krwq @jonsequitur @blackdwarf

### CODE REVIEW

**Comment by jgoshi** [comment]
test Windows_NT x86 Debug Build

File: src/dotnet/commands/dotnet-list/dotnet-list-proj/Program.cs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+using Microsoft.DotNet.Cli.Sln.Internal;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.DotNet.Tools.Common;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+
+namespace Microsoft.DotNet.Tools.List.ProjectsInSolution
+{
+    public class ListProjectsInSolutionCommand
```
Comment (livarcocc) [inline src/dotnet/commands/dotnet-list/dotnet-list-proj/Program.cs]: This is so similar to ListProjectToProjectReferencesCommand, with only slight changes between the two. Are we missing some abstraction?

File: test/dotnet-list-proj.Tests/GivenDotnetListProj.cs [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    {
+        [Theory]
+        [InlineData("--help")]
+        [InlineData("-h")]
+        public void WhenHelpOptionIsPassedItPrintsUsage(string helpArg)
+        {
+            var cmd = new DotnetCommand()
+                .ExecuteWithCapturedOutput($"list projects {helpArg}");
+            cmd.Should().Pass();
+            cmd.StdOut.Should().Contain("Usage");
```
Comment (livarcocc) [inline test/dotnet-list-proj.Tests/GivenDotnetListProj.cs]: I really disagree that we should only check the Usage string. We should check a lot more than this. Why don't we validate the the print out for help is actually the correct one?

File: test/dotnet-list-proj.Tests/GivenDotnetListProj.cs [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        [InlineData("ihave?invalidcharacters.sln")]
+        [InlineData("ihaveinv@lidcharacters.sln")]
+        [InlineData("ihaveinvalid/characters")]
+        [InlineData("ihaveinvalidchar\\acters")]
+        public void WhenNonExistingSolutionIsPassedItPrintsErrorAndUsage(string solutionName)
+        {
+            var cmd = new DotnetCommand()
+                .ExecuteWithCapturedOutput($"list {solutionName} projects");
+            cmd.Should().Fail();
+            cmd.StdErr.Should().Contain("Could not find");
```
Comment (livarcocc) [inline test/dotnet-list-proj.Tests/GivenDotnetListProj.cs]: Again, feels like not enough. Why aren't we doing something like Could not find solutionName?

File: test/dotnet-list-proj.Tests/GivenDotnetListProj.cs [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+            cmd.StdOut.Should().Contain("Usage:");
+        }
+
+        [Fact]
+        public void WhenInvalidSolutionIsPassedItPrintsErrorAndUsage()
+        {
+            var projectDirectory = TestAssetsManager.CreateTestInstance("InvalidSolution")
```
Comment (livarcocc) [inline test/dotnet-list-proj.Tests/GivenDotnetListProj.cs]: Can we use TestAssets.Get instead of TestAssetsManager?

File: test/dotnet-list-proj.Tests/GivenDotnetListProj.cs [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+                                                    .WithLockFiles()
+                                                    .Path;
+
+            var cmd = new DotnetCommand()
+                .WithWorkingDirectory(projectDirectory)
+                .ExecuteWithCapturedOutput("list InvalidSolution.sln projects");
+            cmd.Should().Fail();
+            cmd.StdErr.Should().Contain("Invalid solution ");
```
Comment (livarcocc) [inline test/dotnet-list-proj.Tests/GivenDotnetListProj.cs]: Again, anything else besides invalid solution? Do we say anything as to why the solution is invalid?

File: test/dotnet-list-proj.Tests/GivenDotnetListProj.cs [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+        {
+            var projectDirectory = TestAssetsManager.CreateTestInstance("TestAppWithSlnAndCsprojFiles")
+                                                    .WithLockFiles()
+                                                    .Path;
+
+            var cmd = new DotnetCommand()
+                .WithWorkingDirectory(Path.Combine(projectDirectory, "App"))
+                .ExecuteWithCapturedOutput("list projects");
+            cmd.Should().Fail();
+            cmd.StdErr.Should().Contain("does not exist");
```
Comment (livarcocc) [inline test/dotnet-list-proj.Tests/GivenDotnetListProj.cs]: How about No solution exists in directory <path> or something like that?

File: test/dotnet-list-proj.Tests/GivenDotnetListProj.cs [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+        {
+            var projectDirectory = TestAssetsManager.CreateTestInstance("TestAppWithMultipleSlnFiles")
+                                                    .WithLockFiles()
+                                                    .Path;
+
+            var cmd = new DotnetCommand()
+                .WithWorkingDirectory(projectDirectory)
+                .ExecuteWithCapturedOutput("list projects");
+            cmd.Should().Fail();
+            cmd.StdErr.Should().Contain("more than one");
```
Comment (livarcocc) [inline test/dotnet-list-proj.Tests/GivenDotnetListProj.cs]: Same here.

File: test/dotnet-list-proj.Tests/GivenDotnetListProj.cs [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+            cmd.StdOut.Should().Contain("Usage");
+        }
+
+        [Fact]
+        public void WhenNoProjectReferencesArePresentInTheSolutionItPrintsError()
```
Comment (livarcocc) [inline test/dotnet-list-proj.Tests/GivenDotnetListProj.cs]: I would update the name of this test slightly. You say it should print an error, which gives me an impression that the command should fail. However, we are asserting that the command should pass. Maybe say ItPrintsANoProjectMessage or something like that.

File: test/dotnet-list-proj.Tests/dotnet-list-proj.Tests.csproj [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+  <ItemGroup>
+    <ProjectReference Include="..\Microsoft.DotNet.Tools.Tests.Utilities\Microsoft.DotNet.Tools.Tests.Utilities.csproj" />
+    <ProjectReference Include="..\..\src\Microsoft.DotNet.Cli.Sln.Internal\Microsoft.DotNet.Cli.Sln.Internal.csproj" />
+    <ProjectReference Include="..\..\src\Microsoft.DotNet.Cli.Utils\Microsoft.DotNet.Cli.Utils.csproj" />
+    <ProjectReference Include="..\..\src\Microsoft.DotNet.InternalAbstractions\Microsoft.DotNet.InternalAbstractions.csproj" />
+    <ProjectReference Include="..\..\src\Microsoft.DotNet.TestFramework\Microsoft.DotNet.TestFramework.csproj" />
+  </ItemGroup>
+
+  <ItemGroup Condition=" '$(TargetFramework)' == 'net46' ">
```
Comment (livarcocc) [inline test/dotnet-list-proj.Tests/dotnet-list-proj.Tests.csproj]: We don't target net46 in this project. Why is this here?

File: test/dotnet-list-proj.Tests/dotnet-list-proj.Tests.csproj [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+    <Reference Include="System.Runtime" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.NET.Sdk">
```
Comment (livarcocc) [inline test/dotnet-list-proj.Tests/dotnet-list-proj.Tests.csproj]: Should we hop on the new format for projects for our new projects? I mean, using the SDK attribute instead of the package reference? Putting versions as attributes and removing the imports?

File: src/dotnet/commands/dotnet-list/dotnet-list-proj/Program.cs (hunk 1 above)
Comment (jgoshi) [inline src/dotnet/commands/dotnet-list/dotnet-list-proj/Program.cs]: Ok, I'll try an abstraction

File: test/dotnet-list-proj.Tests/GivenDotnetListProj.cs (hunk 2 above)
Comment (jgoshi) [inline test/dotnet-list-proj.Tests/GivenDotnetListProj.cs]: Ok

Comment (jgoshi) [inline test/dotnet-list-proj.Tests/GivenDotnetListProj.cs]: Although I'm not sure how easy that will be with localization

Comment (livarcocc) [inline test/dotnet-list-proj.Tests/GivenDotnetListProj.cs]: If localization comes in play here, then it will be a problem for a small string or a big one, right? Plus, these tests should run on english.

Comment (jgoshi) [inline test/dotnet-list-proj.Tests/GivenDotnetListProj.cs]: Ok, I'll give this a try.

File: test/dotnet-list-proj.Tests/GivenDotnetListProj.cs (hunk 3 above)
Comment (jgoshi) [inline test/dotnet-list-proj.Tests/GivenDotnetListProj.cs]: Ok

File: test/dotnet-list-proj.Tests/GivenDotnetListProj.cs (hunk 4 above)
Comment (jgoshi) [inline test/dotnet-list-proj.Tests/GivenDotnetListProj.cs]: Ok

File: test/dotnet-list-proj.Tests/GivenDotnetListProj.cs (hunk 5 above)
Comment (jgoshi) [inline test/dotnet-list-proj.Tests/GivenDotnetListProj.cs]: At this point we don't know why the solution is invalid. We just attempted to read and it didn't load. Later on we could try to get smarter about why the solution is invalid. This is tracked by 
https://github.com/dotnet/cli/issues/5036

File: test/dotnet-list-proj.Tests/dotnet-list-proj.Tests.csproj [hunk 11]
```diff
@@ -0,0 +1,49 @@
+﻿<Project ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="$(MSBuildExtensionsPath)\$(MSBuildToolsVersion)\Microsoft.Common.props" />
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
+  
+  <PropertyGroup>
+    <TargetFramework>netcoreapp1.0</TargetFramework>
+    <GenerateRuntimeConfigurationFiles>true</GenerateRuntimeConfigurationFiles>
+    <AssemblyName>dotnet-list-proj.Tests</AssemblyName>
+    <PackageTargetFallback Condition=" '$(TargetFramework)' == 'netcoreapp1.0' ">$(PackageTargetFallback);dotnet5.4;portable-net451+win8</PackageTargetFallback>
+    <DefineConstants>$(DefineConstants);DISABLE_TRACE</DefineConstants>
```
Comment (jgoshi) [inline test/dotnet-list-proj.Tests/dotnet-list-proj.Tests.csproj]: I'll remove the DefineConstants property

File: test/dotnet-list-proj.Tests/dotnet-list-proj.Tests.csproj (hunk 9 above)
Comment (jgoshi) [inline test/dotnet-list-proj.Tests/dotnet-list-proj.Tests.csproj]: Will remove

File: test/dotnet-list-proj.Tests/GivenDotnetListProj.cs [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+        {
+            var projectDirectory = TestAssetsManager.CreateTestInstance("TestAppWithSlnAndExistingCsprojReferences")
+                                                    .WithLockFiles()
+                                                    .Path;
+
+            var cmd = new DotnetCommand()
+                .WithWorkingDirectory(projectDirectory)
+                .ExecuteWithCapturedOutput("list projects");
+            cmd.Should().Pass();
+            cmd.StdOut.Should().Contain("Project reference(s)");
```
Comment (jgoshi) [inline test/dotnet-list-proj.Tests/GivenDotnetListProj.cs]: We don't print that out...

File: src/dotnet/CommandLine/CommandParsingException.cs [hunk 13]
```diff
@@ -7,12 +7,17 @@ namespace Microsoft.DotNet.Cli.CommandLine
...
+            string message,
+            bool isRequireSubCommandMissing = false)
             : base(message)
         {
             Command = command;
+            IsRequireSubCommandMissing = isRequireSubCommandMissing;
         }
 
         public CommandLineApplication Command { get; }
+        public bool IsRequireSubCommandMissing { get; }
```
Comment (krwq) [inline src/dotnet/CommandLine/CommandParsingException.cs]: what's this for? i don't see it used in this class

File: src/dotnet/DotNetTopLevelCommandBase.cs [hunk 14]
```diff
@@ -49,7 +49,11 @@ public int RunCommand(string[] args)
             }
             catch (CommandParsingException e)
             {
-                Reporter.Error.WriteLine(e.Message.Red());
+                string errorMessage = e.IsRequireSubCommandMissing
+                    ? CommonLocalizableStrings.RequiredCommandNotPassed
+                    : e.Message;
```
Comment (krwq) [inline src/dotnet/DotNetTopLevelCommandBase.cs]: Should this logic go to Message in the Exception itself?

File: src/dotnet/commands/dotnet-list/dotnet-list-proj/Program.cs [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+            {
+                _items.Add(slnProject.FilePath);
+            }
+        }
+
+        public string LocalizedErrorMessageNoItemsFound => CommonLocalizableStrings.NoProjectsFound;
+        public IList<string> Items => _items;
+    }
+
+    public class ListProjectsInSolutionCommand : ListSubCommandBase
```
Comment (krwq) [inline src/dotnet/commands/dotnet-list/dotnet-list-proj/Program.cs]: we usually put one class per file

File: src/dotnet/commands/dotnet-list/dotnet-list-p2ps/Program.cs [hunk 16]
```diff
@@ -3,65 +3,51 @@
...
+        protected override IListSubCommand CreateIListSubCommand(string fileOrDirectory)
+        {
+            return new ListProjectToProjectReferences(fileOrDirectory);
+        }
 
-            return app;
+        internal static CommandLineApplication CreateApplication(CommandLineApplication parentApp)
+        {
+            var command = new ListProjectToProjectReferencesCommand();
+            return command.Create(parentApp);
```
Comment (jgoshi) [inline src/dotnet/commands/dotnet-list/dotnet-list-p2ps/Program.cs]: Will consider for the next PR when I implement dotnet remove project

File: src/dotnet/CommandLine/CommandParsingException.cs (hunk 13 above)
Comment (jgoshi) [inline src/dotnet/CommandLine/CommandParsingException.cs]: Used in DotNetTopLevelCommandBase.cs to figure out if we are missing a required sub-command

**Comment by jgoshi** [comment]
@MattGertz

**Comment by livarcocc** [comment]
@srivatsn

**Comment by jgoshi** [comment]
@MattGertz @srivatsn Can one of you please sign off so I can merge this? It's blocking something else I need to do. Thank you.
